Added support for secrets

mhvelplund opened the pull request:
The module now supports an extra parameter `container_secrets`, similar to
`container_envvars`.

Values in this map will be treated as SSM keys and the actual value
injected into the environment of the container will be pulled from SSM.
To work, the SSM keys must also be added to `ssm_paths`.

Example:

```
container_secrets = {
  DB_USER = "arn:aws:ssm:eu-west-1:123456789012:parameter/myapp/dev/db.user"
  DB_PASSWORD = "/myapp/dev/db.password"
}
ssm_paths         = ["myapp/dev"]
```
_Note that non-ARN key references begin with a slash, but the SSM path doesn't._

The example above injects the (decrypted) contents of
`/myapp/dev/db.password` into the env. variable `DB_PASSWORD`.

maartenvanderhoef commented on main.tf:
```diff
@@ -49,6 +49,9 @@ module "iam" {
   fargate_enabled = "${var.fargate_enabled}"
+
+  # The container uses secrets and needs a task execution role to get access to them
+  has_secrets = "${length(keys(var.container_secrets)) > 0}"
```
My suggestion is to rename `has_secrets` into `container_secrets_enabled` as we use this pattern everywhere. I see one issue with this implementation, the moment you populate `container_secrets` from a datasource, some might do this, `cannot compute errors` will occur. For this reason I think it is better to create a module-wide var `container_secrets_enabled` next to the map `container_secrets`.

mhvelplund commented:
Thanks for the feedback. I can't implement and test this before I get back to work after Easter, but then I'll update the PR with all your suggestions and send you a PM in Slack.

mhvelplund commented:
I got around to look at the drift detection you requested in the PR above. Looking at the code, I'm uncertain about how to proceed.

In the lookup lambda, i can see where you get the environment vars, but it doesn't look like javascript's describetaskdefintion returns the secrets as part of the container ...

I use [ecs-deploy](https://github.com/silinternational/ecs-deploy) on my CI server to update the container image when the code changes. The script fetches the running taskdef and replace the image before uploading a new task def.

However, if I run terraform to update something else, it detects a "change" and wants to downgrade to the last task defintion created with Terraform.

I'm not sure if this is related to missing drift detection on secrets, or if it's because `ecs-deploy` removes something that Airship uses to detect the newest image.

mhvelplund commented:
Hmm. Travis built the code (https://travis-ci.org/blinkist/terraform-aws-airship-ecs-service/builds/537776545) but the check result hasn't been fed back. Not sure why.

Anyway, I think I fixed the drift detectionas you requested. I had to finish the docker label support to do so, so I don't think anyone has tried to use that before :)

What do you think about the PR now?

Jamie-BitFlight commented on modules/ecs_container_definition/variables.tf:
```diff
@@ -133,9 +143,19 @@ locals {
     }
   }
 
+  # Secrets aren't trackable by the live_task_loop (thanks AWS!), so we store a hash of them as a synthetic Docker label on the container.
```
nice workaround

Jamie-BitFlight commented on variables.tf:
```diff
@@ -309,6 +314,16 @@ container_envvars defines extra container env vars, list of maps:
 
+variable "container_secrets" {
+  description = <<EOF
+    The environment variables to pass to the container as SSM keys. 
```
Can you mention that the map needs to be set up like:
For secrets manager:

`SOME_ENV_VAR = "arn:aws:secretsmanager:region:aws_account_id:secret:secret_name-AbCdEf"`

For ssm parameter store:

`SOME_ENV_VAR = "arn:aws:ssm:region:aws_account_id:parameter/parameter_name"`

When using SSM within the same region, just the parameter path may be used:
`SOME_ENV_VAR = "parameter_name"`

mhvelplund replied:
Yes.

mhvelplund commented on examples/with_nlb/main.tf:
```diff
@@ -57,40 +58,13 @@ resource "aws_route53_zone" "this" {
-  from_port         = "32768"
-  to_port           = "65535"
-  protocol          = "tcp"
-  cidr_blocks       = ["${formatlist("%s/32",sort(distinct(compact(concat(list(""),data.aws_network_interface.nlb.private_ips)))))}"]
```
I pondered how to do something similar, since getting the NLB's IP is difficult. However, this only seems to work on the first run of the example, and fail on subsequent runs since there are now two interfaces in the subnet.

AFAICT it isn't needed either, since we are running in the default VPC that already has a fitting SG.

Jamie-BitFlight replied:
In subsequent runs what error does it fail with? It should handle having more network interfaces.

mhvelplund replied:
I'm not absolutely sure, but the issue only happened when I added a new service that was dependent on the same NLB. When `data.aws_network_interface.nlb.private_ips` is resolved in the original example folder there is only the one interface belonging to the NLB, but after running, the first service creates a new interface, causing the data element to fail resolution because there are now two interfaces and it only allows one. It might be timing specific, and it might be something resolveable by making the filter tighter, but ultimately it didn't look like it was needed since the default SG only allows traffic from itself, and it would be enough to ad the rule that allows access from the developers own IP. Am I missing something?

mhvelplund commented:
@maartenvanderhoef @Jamie-BitFlight 
Ok, so I added an  example of how to use the new secrets to the example folder. Along the way I added the "sleep.cmd" file that is needed to deploy Airship's ecs-cluster module from Windows. I debated with myself if it was passive agressive, but my coworker convinced me that it might make you remove the "sleep 30" hack from the other module ;)
